src/rules/importBlacklistRule.ts

```diff
@@ -54,9 +88,140 @@ export class Rule extends Lint.Rules.AbstractRule {
+    );
+
     for (const name of findImports(ctx.sourceFile, ImportKind.All)) {
-        if (ctx.options.indexOf(name.text) !== -1) {
-            ctx.addFailure(name.getStart(ctx.sourceFile) + 1, name.end - 1, Rule.FAILURE_STRING);
+        // TODO #3963: Resolve/normalize relative file imports to a canonical path?
+        const importedModule = name.text;
+        const bansForModule = bannedImports[importedModule];
+
+        // Check if at least some imports from this module are banned.
+        if (bansForModule !== undefined) {
+            // If importing this module is totally banned, we can error now,
+            // without determining whether the user is importing the whole
+            // module or named exports.
+            if (bansForModule === true) {
+                ctx.addFailure(
+                    name.getStart(ctx.sourceFile) + 1,
+                    name.end - 1,
+                    Rule.WHOLE_MODULE_FAILURE_STRING,
+                );
+                continue;
+            }
+
+            // Otherwise, find the named imports, if any, and fail if the
+            // user tried to import any of them. We don't have named imports
+            // when the user is importing the whole module, which includes:
+            //
+            // - ImportKind.Require (i.e., `require('module-specifier')`),
+            // - ImportKind.DynamicImport (i.e., `import("module-specifier")`),
+            // - ImportKind.ImportEquals (i.e., `import x = require()`),
+            // - and ImportKind.ImportDeclaration, where there's a full namespace
+            //   import (i.e. `import * as x from "module-specifier"`)
+            //
+            // However, namedImports will be an array when we have one of the
+            // various permutations of `import x, { a, b as c } from "y"`.
+            //
+            // We treat re-exports from other modules the same as attempting to
+            // import the re-exported binding(s), as the re-export is essentially
+            // an import followed by an export, and not treating these as an
+            // import would allow backdoor imports of the banned bindings. So,
+            // our last case is `ImportKind.ExportFrom`, and for that:
+            //
+            // - `export nameForDefault from "module"` isn't part of the ESM
+            // syntax (yet), so we only have to handle two cases below:
+            // `export { x } from "y"` and `export * from "specifier"`.
+            const parentNode = name.parent;
+
+            // Disable strict-boolean-expressions for the next few lines so our &&
+            // checks can help type inference figure out if when don't have undefined.
+            // tslint:disable strict-boolean-expressions
+
+            const importClause = parentNode && isImportDeclaration(parentNode)
+                ? parentNode.importClause
+                : undefined;
+
+            const importsDefaultExport = importClause && Boolean(importClause.name);
```

Comment:
I'm wondering whether there are any benefits of doing `Boolean(importClause.name)` over simple `importClause.name` here and on line 183 if strict-boolean-expressions is disabled?